storage/postgres/query_builder.go

```diff
@@ -0,0 +1,330 @@
+	baseTableName := entity.TableName()
+	dbTags := getDBTags(entity, nil)
+
+	var fieldQueries []string
+
+	if len(criteria) > 0 {
+		qb.sql += " WHERE "
+		for _, option := range criteria {
+			var ttype reflect.Type
+			if dbTags != nil {
+				var err error
+				ttype, err = findTagType(dbTags, option.LeftOp)
+				if err != nil {
+					qb.err = err
+					return qb
+				}
+			}
+			rightOpBindVar, rightOpQueryValue := buildRightOp(option)
+			sqlOperation := translateOperationToSQLEquivalent(option.Operator)
+
+			dbCast := determineCastByType(ttype)
+			clause := fmt.Sprintf("%s.%s%s %s %s", baseTableName, option.LeftOp, dbCast, sqlOperation, rightOpBindVar)
+			if option.Operator.IsNullable() {
+				clause = fmt.Sprintf("(%s OR %s.%s IS NULL)", clause, baseTableName, option.LeftOp)
+			}
+			fieldQueries = append(fieldQueries, clause)
+			qb.queryParams = append(qb.queryParams, rightOpQueryValue)
+		}
+		qb.sql += strings.Join(fieldQueries, " AND ")
+	}
+	return qb
+}
+
+func (qb *QueryBuilder) processResultCriteria(resultQuery []query.Criterion) *QueryBuilder {
+	for _, c := range resultQuery {
+		if c.Type != query.ResultQuery {
+			qb.err = fmt.Errorf("result query is expected, but %s is provided", c.Type)
+			return qb
+		}
+		switch c.LeftOp {
+		case query.OrderBy:
+			if len(c.RightOp) < 1 {
+				qb.err = fmt.Errorf("order by clause expects field and order type, but has none")
+				return qb
+			}
+			if len(c.RightOp) < 2 {
+				qb.err = fmt.Errorf("order by clause (%s) expects order type, but has none", c.RightOp[0])
+				return qb
+			}
+			qb.OrderBy(c.RightOp[0], query.OrderType(c.RightOp[1]))
+		case query.Limit:
+			limit, err := strconv.Atoi(c.RightOp[0])
+			if err != nil {
+				qb.err = err
+				return qb
+			}
+			qb.Limit(limit)
+		}
+	}
+
+	return qb
+}
+
+func (qb *QueryBuilder) expandMultivariateOp() *QueryBuilder {
+	if hasMultiVariateOp(qb.criteria) {
+		var err error
+		// sqlx.In requires question marks(?) instead of positional arguments (the ones pgsql uses) in order to map the list argument to the IN operation
+		if qb.sql, qb.queryParams, err = sqlx.In(qb.sql, qb.queryParams...); err != nil {
+			qb.err = err
+		}
+	}
+	return qb
+}
+
+func (qb *QueryBuilder) orderTypeToSQL(orderType query.OrderType) string {
+	switch orderType {
+	case query.AscOrder:
+		return "ASC"
+	case query.DescOrder:
+		return "DESC"
+	default:
+		qb.err = fmt.Errorf("unsupported order type: %s", string(orderType))
+	}
+	return ""
+}
+
+func validateOrderFields(columns map[string]bool, orderRules ...orderRule) error {
+	fields := make([]string, 0, len(orderRules))
+	for _, or := range orderRules {
+		fields = append(fields, or.field)
+	}
+	return validateFields(columns, fields...)
+}
+
+func validateReturningFields(columns map[string]bool, returningFields ...string) error {
+	if len(returningFields) > 0 {
+		if returningFields[0] == "*" {
+			return nil
+		}
+		return validateFields(columns, returningFields...)
```

Comment:
the error will not be meaningful - enrich it with context about the 'returning` part